### ISSUE TITLE
PNG renderer timeouts under reverse proxy

### ISSUE THREAD

**Issue by frisky-gh** [grafana/grafana#14183]
### What Grafana version are you using?

5.3.4

### What datasource are you using?

influxdb

### What OS are you running grafana on?

Debian stretch

### What did you do?

1. Setup a Apache reverse proxy for a Grafana.  
```
        ProxyPass        /pub/grafana/  http://localhost:14532/
        ProxyPassReverse /pub/grafana/  http://localhost:14532/
```
2. Setup the Grafana under a reverse-proxy.  
```
http_port = 14532
root_url = https://grafana.example.com/pub/grafana/
```
3. Get a PNG image of a dashboard.
https://grafana.example.com/pub/grafana/render/d/000000001/sampledashboard?orgId=1&width=800&height=1200&timeout=120

### What was the expected result?

800x1200 PNG image.

### What happened instead?

> Timeout error. You can set timeout in seconds with &timeout url parameter

/var/log/grafana/grafana.log:
> t=2018-11-26T13:56:00+0900 lvl=info msg=Rendering logger=rendering path="d/000000004/sampledashboard?orgId=2&width=800&height=1200&timeout=120"
> t=2018-11-26T13:56:01+0900 lvl=info msg="Request Completed" logger=context userId=0 orgId=1 uname= method=GET path=/pub/grafana/public/build/grafana.dark.css status=404 remote_addr=127.0.0.1 time_ms=1 size=22207 referer="http://localhost:14532/d/000000001/sampledashboard?orgId=1&width=800&height=1200&timeout=120&render=1"
> t=2018-11-26T13:56:01+0900 lvl=info msg="Request Completed" logger=context userId=0 orgId=1 uname= method=GET path=/pub/grafana/public/build/app.d1f313cb0bbe86ea2d5f.js status=404 remote_addr=127.0.0.1 time_ms=1 size=22207 referer="http://localhost:14532/d/000000001/sampledashboard?orgId=1&width=800&height=1200&timeout=120&render=1"
> t=2018-11-26T13:56:01+0900 lvl=info msg="Request Completed" logger=context userId=0 orgId=1 uname= method=GET path=/pub/grafana/public/build/vendor.d1f313cb0bbe86ea2d5f.js status=404 remote_addr=127.0.0.1 time_ms=6 size=22207 referer="http://localhost:14532/d/000000001/sampledashboard?orgId=1&width=800&height=1200&timeout=120&render=1"

PS info:
```
# ps -ef | grep phantomjs
grafana  29948 28263  0 13:55 ?        00:00:00 /usr/share/grafana/tools/phantomjs/phantomjs --ignore-ssl-errors=true --web-security=true --local-url-access=false --debug=false /usr/share/grafana/tools/phantomjs/render.js url=http://localhost:14532/d/000000001/sampledashboard?orgId=1&width=800&height=1200&timeout=120&render=1 width=800 height=1200 png=/var/lib/grafana/png/Vy5V658Alw47WTwwW9D2.png domain=localhost timeout=120 renderKey=CLnYjm3dO66zyJD376evPunR0pLenTVt
```

### Include raw network request & response: get by opening Chrome Dev Tools (F12, Ctrl+Shift+I on windows, Cmd+Opt+I on Mac), go the network tab.

[Archive 18-11-26 13-49-41.har.txt](https://github.com/grafana/grafana/files/2614141/Archive.18-11-26.13-49-41.har.txt)

**Comment by marefr** [comment]
I can see that you're trying to render a full dashboard and not a single panel. That's not a supported use case in Grafana even though it might work. Please try without reverse proxy first - if that works, then you probably have a configuration error in your reverse proxy setup.

**Comment by frisky-gh** [comment]
I tried some pattern settings.

- I reproduced this issue with "&panelId=1", too.
- With "&timeout=10", I got a PNG image wrote as "Page not found".
![Uploading sampledashboard.png…]()

**Comment by frisky-gh** [comment]
- Without subpath "/pub/grafana/", I got a valid PNG image.
![sampledashboard](https://user-images.githubusercontent.com/33077500/49069158-e90a7800-f26b-11e8-998c-30fd1ee86452.png)

**Comment by frisky-gh** [comment]
By Direct link ( by replacing "/render/d/" -> "/render/d-solo/" ), I got a black PNG image.
![sampledashboard](https://user-images.githubusercontent.com/33077500/49073739-7226ac80-f276-11e8-9156-78214d459660.png)

**Comment by frisky-gh** [comment]
My panel ID was string.
>       "id": "1",
I modified panel ID to number ( "id": "1" -> "id": 1 ), then got a valid image.
![sampledashboard](https://user-images.githubusercontent.com/33077500/49076243-e44dc000-f27b-11e8-83dc-6b476d020b40.png)
Thank you.

**Comment by marefr** [comment]
Did a test run with apache using similar setup as yours and it's not working to render full dashboard due to the extra paths (/pub/grafana/). Some code is needed to make it work:

```diff
diff --git a/public/app/routes/dashboard_loaders.ts b/public/app/routes/dashboard_loaders.ts
index b3e329b..d26575f 100644
--- a/public/app/routes/dashboard_loaders.ts
+++ b/public/app/routes/dashboard_loaders.ts
@@ -33,8 +33,10 @@ export class LoadDashboardCtrl {
     dashboardLoaderSrv.loadDashboard($routeParams.type, $routeParams.slug, $routeParams.uid).then(result => {
       if (result.meta.url) {
         const url = locationUtil.stripBaseFromUrl(result.meta.url);
+        const s = $location.search();
+        const isRender = s && s.render && s.render === '1';
 
-        if (url !== $location.path()) {
+        if (!isRender && url !== $location.path()) {
           // replace url to not create additional history items and then return so that initDashboard below isn't executed multiple times.
           $location.path(url).replace();
           return;
```

**Comment by hegoak64** [comment]
@marefr : can't succesfully apply your patch on grafana6.
dashboard_loaders.ts updated with your patch, restart grafana, but still get the "404 page not found" PNG.

render-solo for panel render is OK, but render for full dashboard still generate the 404 PNG 

Any new idea ?
Thanks

**Comment by marefr** [comment]
Are you trying to render a full dashboard? That's not officially supported. What's your full url when you get a 404?

**Comment by hegoak64** [comment]
Yes I'm trying to render a full dashbaord.
I know it's not "officialy" supported, but it worked with Grafana4 thru reverse proxy
It works also with Grafana6 but without reverse proxy.

The problem occur only in Grafana6 WITH reverse proxy (same apache config)

URL is : https://xxx.intranet/grafana/render/d/000003126/dash-name?orgId=1

Thanks

**Comment by austinjhunt** [comment]
I am having this issue as well, has anyone found a solution?

**Comment by marefr** [comment]
@austinjhunt [this fix](#23706) is included in Grafana v7
